### DISCUSSION TITLE
PI AI camera

### DISCUSSION THREAD

**Discussion by ElectroS3B** [geezacoleman/OpenWeedLocator#160] [General]
I just discovered this https://www.raspberrypi.com/documentation/accessories/ai-camera.html#ai-camera
What do you think, is it a good idea to use it instead of RB-camera-SC0926 Global Shutter?

**Comment by geezacoleman** [comment]
I haven't tested this myself yet, but definitely worth looking into. I don't think the TOPS will be nearly as high as the AI Hats with the Hailo 8 and 8L, but they do claim 30FPS on the [Sony IMX500 spec sheet](https://developer.sony.com/imx500/imx500-key-specifications). It does have an input tensor upper limit of 640 x 480. For reference though, I'm able to get 12 - 13FPS on the Raspberry Pi 5 using a YOLOv8n model at 640 x 480 without any additional hardware but it does get quite warm even with the extra fan.

The Global Shutter has larger pixel sizes 3.4μm × 3.4μm compared to 1.55μm × 1.55μm, so perhaps has better image quality characteristics. We found the v2 camera dealt poorly with high dynamic range regions. The Camera Module 3 is quite nice but the autofocus lens struggles with the vibration on farm equipment.

If you have the opportunity to check it out I would love to see what you find!

**Reply by ElectroS3B** [reply]
Thank you very much for the reply. I will try this camera. I will keep you posted.

**Comment by ElectroS3B** [comment]
I installed an owl with a raspberry ai camera and a 6 relay card. It works in the office and my relays clack when the camera sees green. I would like to see what is happening on the screen. I have no raspberry skills and I do not know how to display what the camera is filming on the screen. I thought that the command line flag "--show-display" is used for this but I must not type it where it is necessary because I have this "command not found". Can you help me please ?

**Reply by geezacoleman** [reply]
That's awesome! Nice work getting it going. Will you be using all 6 relays? If so you'll need to add/change a few things in the config file.

How is the Pi currently set up? Do you have a screen connected? If you can copy your commands/errors in here that might help. The video feed won't display automatically on boot, so you'll need to stop the OWL process that is already running and then restart it with the `--show-display` flag added. But if you just run the commands below in a new terminal window, the new errors should tell you what is already running and how to kill it.

If you want the video feed to appear, you just need to open a terminal window and run: 
1. `cd owl` (to change directory into the owl directory)
2. `python owl.py --show-display`

Let me know how you go. Sounds like you're almost there!

**Reply by ElectroS3B** [reply]
Thank you very much, it works. Yes I want to use 6 relays. For the moment I only change the number of relays and gpio pin for each relay in Day_sensitivity.ini. Now I am trying to understand how green on green works and how use the AI module of my AI raspberry camera.

**Comment by ElectroS3B** [comment]
I have progressed and I managed to convert my trained yolov8n model to imx format. I then imported this model into the imx500 module of the ai camera and the result in the office is interesting. In France in the region where I am it is winter and the frost has destroyed weeds, I cannot test the detection in real conditions, nor train the model. But I find that the test in the office is correct. Here is the result, https://youtu.be/hZUCELyYLxM 
What do you think?
Now the challenge will be to use OWL with this camera by enhancing its imx500 module, I do not know where to start, I do not know how to program but I will look a little. Thanks again for your support.

**Reply by geezacoleman** [reply]
Oh wow this is quite impressive! Thanks for getting this working. You're making me think it could be quite a good option for GoG even if not the best camera quality. It should be supported by the OWL too.

 Are you able to see the framerate of the camera and what resolution it is running?

Integration with the OWL is a little more tricky, but doable. It works with Picamera2 and there are some good documentation. I can add it to the video_manager for automatic detection of the device. If you're happy to wait some weeks I will try integrating this automatically into the OWL using this example: https://github.com/raspberrypi/picamera2/blob/main/examples/imx500/imx500_object_detection_demo_mp.py

Otherwise you can give it a try yourself! You'll just need to replace the video manager file with this example above and add in the detections to the GreenOnGreen class.

**Reply by ElectroS3B** [reply]
Thank you very much. Of course I will wait for the necessary time. I think I am not able to do it myself. To answer the question of resolution and frequency, maybe the desired information is here:


(owl) owl@raspberrypi:~/picamera2/examples/imx500 $ python imx500_object_detection_demo.py --model /home/owl/Desktop/weedai4/network.rpk --fps 30 --bbox-normalization --ignore-dash-labels --bbox-order xy --labels /home/owl/Desktop/weedai4/labels.txt
[26:13:31.878693354] [99231]  INFO Camera camera_manager.cpp:327 libcamera v0.4.0+50-83cb8101
[26:13:31.886164863] [99238]  INFO RPI pisp.cpp:720 libpisp version v1.1.0 e7974a156008 27-01-2025 (21:50:51)
[26:13:31.896121757] [99238]  INFO RPI pisp.cpp:1179 Registered camera /base/axi/pcie@120000/rp1/i2c@88000/imx500@1a to CFE device /dev/media1 and ISP device /dev/media2 using PiSP variant BCM2712_C0

------------------------------------------------------------------------------------------------------------------
NOTE: Loading network firmware onto the IMX500 can take several minutes, please do not close down the application.
------------------------------------------------------------------------------------------------------------------

[26:13:31.935875335] [99231]  INFO Camera camera.cpp:1202 configuring streams: (0) 640x480-XBGR8888 (1) 2028x1520-RGGB_PISP_COMP1
[26:13:31.936166573] [99238]  INFO RPI pisp.cpp:1484 Sensor: /base/axi/pcie@120000/rp1/i2c@88000/imx500@1a - Selected sensor format: 2028x1520-SRGGB10_1X10 - Selected CFE format: 2028x1520-PC1R
Network Firmware Upload: 100%|██████████| 2.88M/2.88M [00:05<00:00, 549kbytes/s]

**Comment by ElectroS3B** [comment]
A new vidéoa new video to show the progress with a yolov8n model trained on young dicotyledons. rpi ai camera with imx500 module 
https://youtu.be/Tm-ZUKJdRe4?si=fKdkR9dVAkQKEm73

**Reply by geezacoleman** [reply]
Nice work! Apologies it's taken longer to get the AI camera integrated than expected. A few additional jobs have popped up.

What approach are you using to get your model converted and uploaded to the camera? Do you have a guide/process that you follow?

**Comment by ElectroS3B** [comment]
No problem, don't apologize and thank you for taking the time to answer me. To train and convert my yolov8n models to imx format, I use google colab and simply this code :
!pip install ultralytics
import ultralytics
ultralytics.checks()

from ultralytics import YOLO

from IPython.display import display, Image
!yolo export model=yolov8n.pt format=imx  # export official model
!yolo export model=/content/runs/detect/train/weights/best.pt format=imx  # export custom trained model

this creates a .zip file named packerOut.zip. I then use the packager tool to get an rpk file
imx500-package -i <path to packerOut.zip> -o <output folder>

this generates a network.rpk file and a label.txt file. I then use this command to launch the camera and this model:

python imx500_object_detection_demo.py --model <path to network.rpk> --fps 17 --bbox-normalization --ignore-dash-labels --bbox-order xy --labels <path to labels.txt>
